libraries/standard/mqtt/src/mqtt.c

```diff
@@ -163,7 +163,29 @@ MQTTStatus_t MQTT_Unsubscribe( MQTTContext_t * const pContext,
 
 MQTTStatus_t MQTT_Disconnect( MQTTContext_t * const pContext )
 {
-    return MQTTSuccess;
+    size_t packetSize;
+    int32_t bytesSent;
+
+    MQTTStatus_t status = MQTT_GetDisconnectPacketSize( &packetSize );
+
+    if( status == MQTTSuccess )
+    {
+        status = MQTT_SerializeDisconnect( &( pContext->networkBuffer ) );
+    }
+
+    if( status == MQTTSuccess )
+    {
+        bytesSent = sendPacket( pContext, packetSize );
+
+        if( bytesSent < 0 )
+        {
+            status = MQTTSendFailed;
+        }
+    }
+
+    pContext->connectStatus = MQTTNotConnected;
```

Comment:
If the send fails, should we keep the connectStatus as-is? Clean disconnect sounds like something that we might recommend that the app dev wrap with a retry.

Reply:
I changed this in response to a comment from @abhidixi11 

> I agree with Gaurav's comment here. Disconnect should always set the connectStatus to disconnected. If the disconnect call fails, there is no point in retrying. Application should then close the transport connection anyway. If application decides to close the transport connection after error and then try reconnect back again, the connect API may fail since the status has not been updated.